align_system/algorithms/outlines_regression_adm.py

```diff
@@ -0,0 +1,312 @@
+        # machine from looping indefinitely in some cases, see:
+        # https://github.com/outlines-dev/outlines/issues/690#issuecomment-2102291934
+        outcome_generator = outlines.generate.json(
+            self.model,
+            outcome_prediction_json_schema(),
+            whitespace_pattern=r"[ ]?")
+
+        outcome_dialog_texts = [self.dialog_to_prompt(d) for d in
+            itertools.chain(outcome_dialogs)]
+
+        log.info("[bold]*OUTCOMES PREDICTION DIALOG PROMPT*[/bold]",
+                 extra={"markup": True})
+        log.info(outcome_dialog_texts[0])
+
+        # List of {predicted_outcomes:}, one of each choice in order of choices
+        predicted_outcomes = outcome_generator(outcome_dialog_texts)
+
+        log.info("[bold]*OUTCOME PREDICTION RESPONSE*[/bold]",
+                 extra={"markup": True})
+        log.info(predicted_outcomes, extra={"highlighter": JSON_HIGHLIGHTER})
+
+        return predicted_outcomes
+
+    # Predicts kdma scores associated with each choice
+    # Outputs a list of ADM responses and a corresponding list of response keys:
+    #   kdma_score_responses = [{score:int, reasoning:str}, ...]
+    #   reponse_keys = [{kdma:str, choice:str}, ...]
+    def predict_kdma_scores(self, scenario_description, choices, target_kdmas, predicted_outcomes=None):
+        kdma_dialogs = []
+        response_keys = []
+        # loop over target kdmas
+        for target_kdma in target_kdmas:
+            kdma_score_sys_prompt = kdma_score_prediction_system_prompt(target_kdma['name'], target_kdma['description'])
+            # loop over choices
+            for choice_idx in range(len(choices)):
+                choice = choices[choice_idx]
+                if predicted_outcomes:
+                    outcome = predicted_outcomes[choice_idx]['predicted_outcome']
+                else:
+                    outcome = None
+                predict_kdma_prompt = kdma_score_prediction_prompt(scenario_description, choice, outcome, target_kdma['name'])
+                kdma_dialogs.append([{'role': 'system', 'content': kdma_score_sys_prompt},
+                            {'role': 'user', 'content': predict_kdma_prompt}])
+                response_keys.append({'kdma':target_kdma['kdma'], 'choice':choice})
+
+        # Need to set the whitespace_pattern to prevent the state
+        # machine from looping indefinitely in some cases, see:
+        # https://github.com/outlines-dev/outlines/issues/690#issuecomment-2102291934
+        kdma_score_generator = outlines.generate.json(
+            self.model,
+            kdma_score_prediction_json_schema(),
+            whitespace_pattern=r"[ ]?")
+
+        kdma_dialog_texts = [self.dialog_to_prompt(d) for d in
+            itertools.chain(kdma_dialogs)]
+
+        log.info("[bold]*KDMA SCORE PREDICTION DIALOG PROMPT*[/bold]",
+                 extra={"markup": True})
+        log.info(kdma_dialog_texts[0])
+
+        # List of {score:int, reasoning:str}
```

Comment:
Thanks for pointing this out. The original implementation did have the LLM provide reasoning first so I'll leave this order for now, but this will be good to reconsider when we think about ways to improve performance